### PR TITLE
xds: Improvements to the xDS delegate and its KV store test

### LINKED ISSUE #26974: `IpVersions/KeyValueStoreXdsDelegateIntegrationTest.BasicSuccess` is flakey
I have seen this a few times, but infrequently, and only recently

```console

[       OK ] IpVersions/KeyValueStoreXdsDelegateIntegrationTest.BasicSuccess/3 (268 ms)
[ RUN      ] IpVersions/KeyValueStoreXdsDelegateIntegrationTest.BasicSuccess/4
./test/integration/server.h:445: Failure
Value of: TestUtility::waitForCounterGe(statStore(), name, value, time_system_, timeout)
  0x1b977d4: testing::Test::Run()
  0x1b98924: testing::TestInfo::Run()
... Google Test internal frames ...

contrib/config/test/kv_store_xds_delegate_integration_test.cc:360: Failure
Expected equality of these values:
  2
  test_server_->counter("xds.kv_store.load_success")->value()
    Which is: 1
Stack trace:
  0x1b97934: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x1b977d4: testing::Test::Run()
  0x1b98924: testing::TestInfo::Run()
... Google Test internal frames ...

contrib/config/test/kv_store_xds_delegate_integration_test.cc:367: Failure
Expected equality of these values:
  "saz"
  getRuntimeKey("baz")
    Which is: ""
Stack trace:
  0x1b97934: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x1b977d4: testing::Test::Run()
  0x1b98924: testing::TestInfo::Run()
... Google Test internal frames ...

unknown file: Failure
C++ exception with description "cannot bind '[::1]:45237': Address already in use" thrown in the test body.
[external/com_google_absl/absl/flags/internal/flag.cc : 115]
```

https://dev.azure.com/cncf/envoy/_build/results?buildId=135269&view=logs&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=67a76341-37f4-5b49-db38-acbb8a59af82&l=213

for some reason im not seeing any trace of this in azp analytics, not sure why

The issue would appear to be running out of (or trying to reuse currently active) ports

### PR DESCRIPTION
**Pull request by abeyad**
This commit contains a few improvements to the xDS delegate:

 1. Fixes a bug where the legacy GrpcMuxImpl did not keep track of xDS delegate initialization per type url, so it could happen that some type urls would not get initialized from the xDS delegate if the xDS management server was down.
 2. Adds the ability for the FakeUpstream to optionally differ listener initialization and leave it in manual control of the programmer, instead of initializing the listener for incoming connections by default in the constructor. This is useful if you want to start a FakeUpstream and claim a listener port, but not have the FakeUpstream listen for incoming connections yet until a deferred time. If using deferred initialization, the programmer must explicitly call the `FakeUpstream::initializeServer()` function.
 3. Fixes flakiness in the KeyValueStoreXdsDelegateIntegrationTest by creating, but not initializing, the FakeUpstream (using the functionality described in 2). This ensures that the KV store is always used when restarting the Envoy server, then initializing the FakeUpstream, then ensuring the management server's xDS is used when connectivity is re-established.
 4. Using the FakeUpstream with deferred initialization also allows us to get around the problem of using the same port and getting occasional "address already in use" errors, while still enabling the FakeUpstream to start in a state where it isn't listening for incoming connections, so that the xDS delegate / KV store kicks in.

Testing: `bazelisk test  --runs_per_test=100 --test_output=streamed --test_arg="-l debug" --cache_test_results=no //contrib/config/test:kv_store_xds_delegate_integration_test "--test_arg=--gtest_filter=*IpVersions/KeyValueStoreXdsDelegateIntegrationTest.BasicSuccess/*"`

Fixes #26974

### CODE REVIEW

**Review by adisuissa — COMMENTED** [review]
Thanks!
Core code changes LGTM.
Left a note about the deferred initialization, that I'm not certain if it is needed.

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (abeyad) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:263]: when you say explicit, do you mean adding a `removeFakeUpstream()` function to BaseIntegrationTest, as well as using `addFakeUpstream` and adding a param to specify deferred initialization?

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (abeyad) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:266]: I can't wait until the test is running, b/c by that time, the Bootstrap config is already created (with the xds cluster endpoints with their port numbers), so I need to know the listening port number of the FakeUpstream prior to Envoy recreation.  However, I don't want the FakeUpstream to start listening on that port until after the Envoy server thinks the FakeUpstream is down and fetches xDS from the KV store instead. After fetching from the KV store, I want to have the FakeUpstream listener tied to the server so it starts listening for incoming connections and gets the config from the xDS FakeUpstream.

This was the only way I could get the tests to do what I wanted.

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (adisuissa) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:263]: I think that the `fake_upstreams_` were expected to be used for backend hosts, and somehow many of the integration tests are using them also for xDS servers.
It seems that the case here needs more control over those explicitly, so it may be better to use 2 explicit "upstreams" members: `rtds_upstream_`, and `sds_upstream_` each having the FakeUpstream, its connection, etc.
That way the removal/addition to the fake_upstreams_ is avoided.

An example where the current use of `fake_upstreams_` could backfire is if there was also data-plane cluster, which was added after the 2 (rtds, sds) upstreams.

No AI here for now, just something to think about for the future.

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (adisuissa) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:266]: What I think I'm misunderstanding, is why an Envoy (test-server) takedown requires restarting of the upstreams (the xDS servers).
They can keep running, using the same ports as before, and wait for new connections. The restarted Envoy's bootstrap will need to be updated with the previous ports.

**Review by adisuissa — APPROVED** [review]
Thanks!
Left a couple of things to think about, but I don't think they should block this PR.
LGTM

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (abeyad) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:263]: ah ok I see what you mean now.  That's a fair point, though I think we'll have to rework some logic in BaseIntegrationTest.  Other xDS tests follow the same pattern of using fake_upstreams_ for the xDS servers too, e.g. https://github.com/envoyproxy/envoy/blob/e2a40eaff8baf35b269a936526512417033b916a/test/integration/xds_integration_test.cc#L869-L878

I agree - I think it's worth trying to figure out a better way for this in a follow up PR.

**Comment by abeyad** [comment]
added @lizan for cross-company review

File: contrib/config/test/kv_store_xds_delegate_integration_test.cc
Comment (adisuissa) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:266]: I thought (maybe misremembering) that the delegate will provide the data in case the xDS server disconnects immediately, but I guess that not responding is an easier way to trigger the use of the delegate.

Comment (abeyad) [reply contrib/config/test/kv_store_xds_delegate_integration_test.cc:266]: yeah, unfortunately that wouldn't trigger the delegate... the xDS server needs to not respond at Envoy server startup to trigger the delegate

**Review by lizan — APPROVED** [review]
LGTM, defer to @adisuissa for merging.

**Pull request merged**
